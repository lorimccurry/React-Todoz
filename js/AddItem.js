import React from 'react'

class AddItem extends React.Component {
  constructor () {
    super()

    this.state = {
      itemInput: ''
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange (e) {
    this.setState({itemInput: e.target.value})
  }

  handleSubmit (e) {
    e.preventDefault()
    this.props.onSubmit(this.state.itemInput)
    this.setState({itemInput: ''})
  }

  render () {
    return (
      <form onSubmit={this.handleSubmit}>
        <fieldset>
          <legend>Gimme tha todoz:</legend>
          <input
            type='text'
            placeholder='Get coffee'
            value={this.state.itemInput}
            onChange={this.handleChange}
            autoFocus
          />
        </fieldset>
      </form>
    )
  }
}

const { string, func } = React.PropTypes
AddItem.propTypes = {
  itemInput: string,
  onSubmit: func
}

export default AddItem
