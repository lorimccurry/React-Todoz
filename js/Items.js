import React from 'react'
import Item from './Item'
import EditItem from './EditItem'

function Items (props) {
  const items = props.items
  if (!items) {
    return null
  }
  function renderListItems (items, props) {
    return items.map((item) => {
      if (props.editID === item.id) {
        return <EditItem
          key={item.id}
          {...item}
          handleEditID={(itemID) => props.handleEditID(itemID)}
          handleUpdateItemInput={(e, itemID) => props.handleUpdateItemInput(e, itemID)}
          onKeyDown={(e, itemID) => props.onKeyDown(e, itemID)}
        />
      } else {
        return <Item
          key={item.id}
          {...item}
          handleEditID={(itemID) => props.handleEditID(itemID)}
          handleUpdateItemInput={(e, itemID) => props.handleUpdateItemInput(e, itemID)}
          onKeyDown={(e, itemID) => props.onKeyDown(e, itemID)}
          handleDeleteItem={(itemID) => props.handleDeleteItem(itemID)}
        />
      }
    })
  }

  return (
    <ul>
      {renderListItems(items, props)}
    </ul>
  )
}

const { arrayOf, shape, string, bool, func } = React.PropTypes
Items.propTypes = {
  items: arrayOf(shape({
    text: string,
    id: string,
    completed: bool
  })),
  editID: string,
  handleDeleteItem: func,
  handleEditID: func,
  handleUpdateItemInput: func,
  onKeyDown: func
}

export default Items
