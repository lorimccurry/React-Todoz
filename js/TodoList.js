import React from 'react'
import shortid from 'shortid'
import update from 'immutability-helper'
import AddItem from './AddItem'
import Items from './Items'

class TodoList extends React.Component {
  constructor () {
    super()

    this.state = {
      items: [],
      editID: ''
    }
    this.handleAddItem = this.handleAddItem.bind(this)
  }

  handleAddItem (itemInput) {
    const newTodo = this.createNewTodo(itemInput)
    const method = '$push'
    const newItemsState = this.immutableUpdateState(this.state.items, [newTodo], method)
    this.setState({items: newItemsState})
  }

  handleDeleteItem (deleteId) {
    const updatedItems = this.state.items.filter((item) => { return item.id !== deleteId })
    const method = '$set'
    const newItemsState = this.immutableUpdateState(this.state.items, updatedItems, method)
    this.setState({items: newItemsState})
  }

  handleReturnKeySubmit (e, itemID) {
    if (e.keyCode === 13) {
      e.preventDefault()
      e.target.type === 'text' ? this.setEditID(itemID) : this.handleUpdateItemInput(e, itemID)
    } else {
      return
    }
  }

  handleUpdateItemInput (e, itemID) {
    const inputType = e.target.type
    const itemPropertyKey = e.target.name
    const eventValue = e.target.value
    const method = '$set'
    const currentItems = this.state.items

    const mutatedItemArr = currentItems.map((item, index) => {
      if (item.id === itemID) {
        inputType === 'checkbox' ? item[itemPropertyKey] = !item[itemPropertyKey] : item[itemPropertyKey] = eventValue
        this.immutableUpdateState(currentItems[index], [item], method)
        return item
      } else {
        return item
      }
    })
    this.setState({items: mutatedItemArr})
  }

  createNewTodo (itemInput) {
    return {
      completed: false,
      id: shortid.generate(),
      text: itemInput
    }
  }

  immutableUpdateState (currentState, updatedState, method) {
    return update(currentState, {[method]: updatedState})
  }

  setEditID (editID) {
    this.setState({editID: editID})
  }

  renderListDisplay () {
    if (this.state.items.length > 0) {
      return <Items
        items={this.state.items}
        editID={this.state.editID}
        handleDeleteItem={(deleteId) => this.handleDeleteItem(deleteId)}
        handleEditID={(editID) => this.setEditID(editID)}
        handleUpdateItemInput={(e, itemID) => this.handleUpdateItemInput(e, itemID)}
        onKeyDown={(e, itemID) => this.handleReturnKeySubmit(e, itemID)}
        />
    } else {
      return <p>Add some todoz!</p>
    }
  }

  render () {
    return (
      <div>
        <AddItem onSubmit={this.handleAddItem} />
        {this.renderListDisplay()}
      </div>
    )
  }
}

const { string, array } = React.PropTypes
TodoList.propTypes = {
  items: array,
  editID: string
}

export default TodoList
