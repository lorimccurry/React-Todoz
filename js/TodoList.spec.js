import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import TodoList from './TodoList'
import Items from './Items'
import AddItem from './AddItem'

describe('TodoList', () => {
  it('TodoList snapshot test', () => {
    const component = shallow(<TodoList />)
    const tree = shallowToJson(component)
    expect(tree).toMatchSnapshot()
  })
})

describe('TodoList add item', () => {
  it('renders AddItem component', () => {
    const component = shallow(<TodoList />)
    expect(component.find(AddItem).length).toEqual(1)
  })

  it('tests another way to test rendering AddItem component', () => {
    const component = shallow(<TodoList />)
    expect(component.containsMatchingElement(AddItem)).toEqual(true)
  })

  it('should start with an empty list', () => {
    const component = shallow(<TodoList />)
    expect(component.state('items')).toEqual([])
  })

  it('adds item to the items state', () => {
    const expected = [
      {
        completed: false,
        id: expect.any(String),
        text: 'get beer'
      }
    ]
    const component = shallow(<TodoList />)
    component.instance().handleAddItem('get beer')
    expect(component.state('items')).toEqual(expected)
  })

  it('adds items to the items state', () => {
    const component = shallow(<TodoList />)
    component.instance().handleAddItem('get beer')
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    expect(component.state('items').length).toEqual(3)
  })

  it('renders instruction message when no items', () => {
    const component = shallow(<TodoList />)
    expect(component.state('items').length).toEqual(0)
    expect(component.contains(<p>Add some todoz!</p>)).toEqual(true)
  })

  it('does not render Items when no items', () => {
    const component = shallow(<TodoList />)
    expect(component.find(Items).length).toEqual(0)
  })

  it('renders Items when item added', () => {
    const component = shallow(<TodoList />)
    component.instance().handleAddItem('get beer')
    expect(component.state('items').length).toEqual(1)
    expect(component.containsMatchingElement(<Items />)).toEqual(true)
  })

  it('passes handleAddItem to AddItem', () => {
    const component = shallow(<TodoList />)
    const addItem = component.find(AddItem)
    const handleAddItem = component.instance().handleAddItem
    expect(addItem.prop('onSubmit')).toEqual(handleAddItem)
  })

  it('passes a bound handleAddItem function to AddItem', () => {
    const component = shallow(<TodoList />)
    const addItem = component.find(AddItem)
    const expected = [
      {
        completed: false,
        id: expect.any(String),
        text: 'get beer'
      }
    ]
    addItem.prop('onSubmit')('get beer')
    expect(component.state('items')).toEqual(expected)
  })
})

describe('TodoList show Item in Items', () => {
  it('renders added items', () => {
    const component = mount(<TodoList />)   
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')
    expect(component.find('li').length).toEqual(3)
  }) 
})

describe('TodoList deletes', () => {
  it('handleDeleteItem deletes item from state and view', () => {
    const component = mount(<TodoList />)
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')
    expect(component.state('items').length).toEqual(3)
    expect(component.find('li').length).toEqual(3)

    const itemID = component.state('items')[0].id
    component.instance().handleDeleteItem(itemID)
    expect(component.state('items').length).toEqual(2)
    expect(component.find('li').length).toEqual(2)
  })

  it('deletes and renders the correct item when delete button clicked', () => {
    const component = mount(<TodoList />)
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')
    expect(component.find('li').length).toEqual(3)

   const deleteItem = component.find('Item').first()
    const deleteItemID = deleteItem.prop('id')
    const origItemsArr = component.find('Items').prop('items')

    function isMatchingItem(item) {
      return item.id === deleteItemID 
    }
    const initialMatchingItemsArr = origItemsArr.filter(isMatchingItem)
    expect(initialMatchingItemsArr.length).toEqual(1)

    const deleteBtn = deleteItem.find('.delete-btn')
    deleteBtn.simulate('click')

    const updatedItemsArr = component.find('Items').prop('items')
    const finalMatchingItemsArr = updatedItemsArr.filter(isMatchingItem)

    expect(component.find('li').length).toEqual(2)
    expect(finalMatchingItemsArr.length).toEqual(0)
  })
})

describe('TodoList Edit', () => {
  it('updates EditItem state and view', () => {
    const component = mount(<TodoList />)
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')

    const editBtn = component.find('.edit-btn').first()
    editBtn.simulate('click')

    const updateInput = component.find('.update-input')
    expect(updateInput.prop('value')).toEqual('eat')

    const event = {target: {type: 'text', name: 'text', value: 'codez'}}
    updateInput.simulate('change', event)

    const itemStateText = component.state('items')[0].text
    const itemValue = updateInput.prop('value')
    expect(itemStateText).toEqual('codez')
    expect(itemValue).toEqual('codez')
  })

  it('updates Item text state and view', () => {
    const component = mount(<TodoList />)
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')

    let item = component.find('Item').first()
    const itemStateText = component.state('items')[0].text
    const itemValue = item.find('.item-text').text()
    expect(itemStateText).toEqual('eat')
    expect(itemValue).toEqual('eat')

    const editBtn = item.find('.edit-btn')
    editBtn.simulate('click')

    const editItem = component.find('EditItem')
    const updateInput = editItem.find('.update-input')
    const updateBtn = editItem.find('.update-btn')
    const updateEvent = {target: {type: 'text', name: 'text', value: 'codez'}}
    updateInput.simulate('change', updateEvent)
    updateBtn.simulate('click')

    item = component.find('Item').first()
    const updatedItemStateText = component.state('items')[0].text
    const updatedItemValue = item.find('.item-text').text()
    expect(updatedItemStateText).toEqual('codez')
    expect(updatedItemValue).toEqual('codez')
  })

  it('updates Item checkbox state and view', () => {
    const component = mount(<TodoList />)
    component.instance().handleAddItem('eat')
    component.instance().handleAddItem('sleep')
    component.instance().handleAddItem('live')

    let item = component.find('Item').first()
    const itemCheckbox = item.find('input[name="completed"]')
    const itemCheckboxState = component.state('items')[0].completed
    const itemCheckboxChecked = itemCheckbox.prop('checked')
    expect(itemCheckboxState).toEqual(false)
    expect(itemCheckboxChecked).toEqual(false)

    itemCheckbox.simulate('change')
    const updatedItemCheckboxState = component.state('items')[0].completed
    const updatedItemCheckboxChecked = itemCheckbox.prop('checked')
    expect(updatedItemCheckboxState).toEqual(true)
    expect(updatedItemCheckboxChecked).toEqual(true)
  })
})
