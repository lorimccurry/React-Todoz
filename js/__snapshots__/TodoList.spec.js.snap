// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList TodoList snapshot test 1`] = `
<div>
  <AddItem
    onSubmit={[Function]}
  />
  <p>
    Add some todoz!
  </p>
</div>
`;
